cheri-c-programming.tex

```diff
@@ -1480,6 +1507,151 @@ \subsection{Handling capability-related signals}
+<source>:2:17: note: insert cast to intptr_t to silence this warning
+1 warning generated.
+\end{compilerwarning}
+
+\noindent
+In some cases this warning can be a false-positive.
+For example, it is common for C callback APIs take a \cvoidstar data argument that is passed to the callback.
+If this value is in fact an integer constant, the warning can be silenced by casting to \cuintptrt first:
+\TikzListingHighlightStartEnd[green]{SilenceProv}
+\begin{clisting}[frame=single,numbers=left]
+void invoke_cb(void(*cb)(void*), void*);
+void callback(void* arg);
+void false_positive_example(int callback_data) {
+  invoke_cb(&callback, (void*)callback_data); // warning
+  invoke_cb(&callback, (void*)£\vcpgfmark{StartSilenceProv}£(uintptr_t)£\vcpgfmark{EndSilenceProv}£callback_data); // no warning
+}
+\end{clisting}
+\begin{compilerwarning}
+<source>:4:24: warning: cast from provenance-free integer type to pointer type will give pointer that can not be dereferenced [-Wcheri-capability-misuse]
+  invoke_cb(&callback, (void*)callback_data); // warning
+                       ^
+<source>:4:24: note: insert cast to intptr_t to silence this warning
+<source>:15:24: warning: cast to 'void * __capability' from smaller integer type 'int' [-Wint-to-void-pointer-cast]
+  invoke_cb(&callback, (void*)callback_data); // warning
+                       ^
+2 warnings generated.
+\end{compilerwarning}
+
+\subsection{Ambiguous provenance}
+For arithmetic and bitwise binary operations between \cuintptrt/\cintptrt, the compiler can generally infer which side of the expression should be used as the provenance (and bounds) source.
+However, as noted in \cref{sec:ambiguous-provenance}, there are cases that are ambiguous as far as the compiler is concerned.
+
+Consider for example a structure that holds a pointer and a small amount of flags.
+In this case the pointer is known to be aligned to at least 8 bytes, so the programmer uses the lowest 3 bits to store additional data:
+\begin{clisting}[frame=single,numbers=left]
+typedef struct { uintptr_t data; } pointer_and_flags;
+void set_ptr(pointer_and_flags* p, void* value) {
+    p->data = (p->data & (uintptr_t)7) | (uintptr_t)(value);
+}
+void set_flags(pointer_and_flags* p, unsigned flags) {
+    p->data = p->data | (flags & 7);
+}
+\end{clisting}
+\begin{compilerwarning}
+<source>:3:40: warning: binary expression on capability types '__uintcap_t' and 'uintptr_t' (aka '__uintcap_t'); it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side [-Wcheri-provenance]
+    p->data = (p->data & (uintptr_t)7) | (uintptr_t)(value);
+              ~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~
+1 warning generated.
+\end{compilerwarning}
+
+\noindent
+Unlike the compiler, the programmer knows that inside \cfunc{set\_ptr} capability metadata should always be taken from the \ccode{value} argument.
+The suggested fix for this problem is to insert a cast to an integer type:
```

Comment:
Maybe "fix for this problem is to cast the non-pointer argument to an integer type" or something similar to be explicit about which of the two expressions the cast should be added to.

Reply:
In some regards I'd like the fix to be adding a `__noprovinance` qualifier to the non-pointer `uintptr_t`, but I don't think we have that. My thinking is that finding a "comparable" integer type isn't necessarily trivial (at least in portable code).

Reply:
We do have that, but usually the cast is less verbose.